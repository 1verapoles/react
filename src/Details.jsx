import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types';
import { useParams } from "react-router-dom";
import Spinner from './Spinner'
import './Card.css'
import imgDef from './assets/images/comp.jpg'
const API_KEY = "2aedd492fa3c4e97b41ac78483bc25e1"

function Details() {
  const [isLoading, setIsLoading] = useState(false)
  const [card, setCard] = useState({})
  const { id } = useParams()
  useEffect(() => {
    let data = localStorage.getItem(id)
    if (data) { data = JSON.parse(data); setCard(data); return; }
    const fetchData = (url) => {
      setIsLoading(true)
      fetch(url)
        .then(response => response.json())
        .then(json => { setCard(json.articles[0]); localStorage.setItem(id, JSON.stringify(json.articles[0])); setIsLoading(false) })
        .catch(error => { console.log(error); setIsLoading(false) })
    }
    fetchData(`https://newsapi.org/v2/everything?qInTitle=${id}&apiKey=${API_KEY}`)
  }, [])

  const { title, description, author, publishedAt, urlToImage, url } = card
  let date = new Date(publishedAt).toLocaleString().split(',')[0];
  return (
    <div className="container d-flex justify-content-center">
      {isLoading && <Spinner />}
      {title && <div className="card mt-5 card2" >
        <img src={urlToImage || imgDef} className="card-img-top" alt={title} />
        <div className="card-body card-body2">
          <h5 className="card-title">{title}</h5>
          <p className="card-text">{description}</p>
        </div>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">Автор: {author}</li>
          <li className="list-group-item">Дата публикации: {date}</li>
        </ul>
        <div className="card-body">
          <a href={url} target="_blank" rel="noreferrer" className="card-link">Подробнее</a>
        </div>
      </div>}
    </div>
  )
}

Details.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  author: PropTypes.string,
  publishedAt: PropTypes.string,
  url: PropTypes.string,
  urlToImage: PropTypes.string
};

export default Details
