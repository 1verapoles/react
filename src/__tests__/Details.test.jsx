import React from 'react';
import { BrowserRouter } from "react-router-dom";
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../store';
import Details from '../Details';

const card = { title: "new", description: "desc", author: "Marry", publishedAt: "2021-09-03T13:30:41Z", urlToImage: "https://i.pinimg.com/564x/ff/a8/f7/ffa8f756bf1b3cbb28606b89b09da658.jpg", url: "https://i.pinimg.com/564x/ff/a8/f7/ffa8f756bf1b3cbb28606b89b09da658.jpg" }

describe('Testing Details component', () => {

  const unmockedFetch = global.fetch

  beforeAll(() => {
    global.fetch = () =>
      Promise.resolve({
        json: () => Promise.resolve({ articles: [card] }),
      })
  })

  afterAll(() => {
    global.fetch = unmockedFetch
  })
  it('renders without crashing', () => {

    const { asFragment } = render(<Provider store={store}><BrowserRouter><Details /></BrowserRouter></Provider>);

    expect(asFragment()).toMatchSnapshot();

  });
});
