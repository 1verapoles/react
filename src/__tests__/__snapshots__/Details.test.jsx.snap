// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Details component renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="container d-flex justify-content-center"
  >
    <div
      class="d-flex justify-content-center"
    >
      <div
        class="spinner-border text-primary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
